shared-libs/cht-datasource/src/libs/data-context.ts

```diff
@@ -39,3 +39,30 @@ export const adapt = <T>(
   assertRemoteDataContext(context);
   return remote(context);
 };
+
+interface PaginationArgs {
+  limit: number;
+  skip: number;
+}
+
+type FetchFunctionArgs<TFields extends Record<string, unknown>> = PaginationArgs & TFields;
+
+/** @internal */
+export const getDocumentStream = async function* <T, TFields extends Record<string, unknown>>(
+  fetchFunction: (args: FetchFunctionArgs<TFields>) => Promise<T[]>,
+  fetchFunctionArgs: FetchFunctionArgs<TFields>
+): AsyncGenerator<T, void> {
+  const { limit } = fetchFunctionArgs;
+  let { skip } = fetchFunctionArgs;
+  const hasMoreResults = () => skip && skip % limit === 0;
```

Comment:
Thinking out loud, and definitely not necessary at this stage, but we could re-use the numbers in CouchDB's response to know if we've hit the limit here. That would save us one extra query with an empty page in the rare scenario where `total_rows % 100 === 0`
I shared what CouchDB's response looks like in [this comment](https://github.com/medic/cht-core/pull/9266#discussion_r1697035442). For this particular query, the `hasMoreResults` function could be
```ts
const hasMoreResults = () => response.total_rows > (response.offset + response.rows.length) 
```

Again, you don't need to change anything right now, the current logic is good enough for our needs today